gulpfile.js

```diff
@@ -0,0 +1,33 @@
+
+// used by travis-ci to check the hiera json files
+
+var gulp = require("gulp");
+var jsonlint = require("gulp-jsonlint");
+var expect = require('gulp-expect-file');
+var gutil = require('gulp-util');
+
+// custom reporter that should throw exception on lint fail
+var myFailReporter = function (file) {
```

Comment:
We used to have a styleguide page up somewhere but I can't find it anymore.

Here's what I could scrape together:
```
Braces

Opening braces go on the same line as the statement.

Right:

if (true) {
    debug.log('winning');
}
Wrong:

if (true)
{
    debug.log('losing');
}
Also notice the use of whitespace before and after the condition statement.

Callbacks

Functions often return data by using callbacks. All API functions that return data do this through a callback function and it's recommended you use them.

Conditions

Any non-trivial conditions should be assigned to a descriptive variable. This improves code readability.

Right:

var isAuthorized = (user.isAdmin() || user.isModerator());
if (isAuthorized) {
    debug.log('winning');
}
Wrong:

if (user.isAdmin() || user.isModerator()) {
    debug.log('losing');
}
Constants

Constants should be declared as regular variables or static class properties, using all capital letters.

Right:

var SECOND = 1 * 1000;
Wrong:

const second = 1 * 1000;
Documentation

Use JSDoc to document functions.

/**
 * Convert a number to a boolean
 * @param {Number} num The number to convert
 * @return {Boolean} The number in boolean form
 */
var convertToBoolean = function(num) {
   return num !== 0;
};
Editors

Any editor can be used. However, it's recommended to use an editor that respects these guidelines (tabs vs spaces, etc.) and a little syntax highlighting never hurt anyone. Frequently used editors in the team are <%= link_to "Textmate", "http://macromates.com/", :class => "wl-regular-link" %>, <%= link_to "Sublime", "http://www.sublimetext.com/", :class => "wl-regular-link" %> and <%= link_to "Aptana Studio", "http://www.aptana.com/", :class => "wl-regular-link" %>.

Equality operator

Don't take a guess at what you're comparing, use triple equality operators at all times.

Right:

var a = 0;
if (a === '') {
    debug.log('winning');
}
Wrong:

var a = 0;
if (a == '') {
    debug.log('losing');
}
Event handling

As of jQuery 1.7 the .live(), .die(), .bind() and .unbind() methods are deprecated, instead use .on() to attach event handlers and .off() to remove them.

var changeEmail = function() {
    ...
};

$('#accountpreferences_email').off('change', changeEmail);
// or
$rootel.off('change', '#accountpreferences_email', changeEmail);

...

$('#accountpreferences_email').on('change', changeEmail);
// or
$rootel.on('change', '#accountpreferences_email', changeEmail);
Use events to send out updates to other widgets.

var worldTitle = 'OAE Core Team';
$(window).trigger('updatedTitle.worldsettings.sakai', worldTitle);

$(window).on('updatedTitle.worldsettings.sakai', function(ev, title) {
    debug.log(title); // "OAE Core Team"
});
Extending prototypes

Don't extend the prototypes of any objects.

Right:

var a = [];
if (!a.length) {
    debug.log('winning');
}
Wrong:

Array.prototype.empty = function() {
    return !this.length;
}

var a = [];
if (a.empty()) {
    debug.log('losing');
}
Function declaration

Functions should be assigned to a variable instead of declared inline.

Right:

var myFunction = function() {
    console.log('super cool function!');
};
Wrong:

function myFunction() {
    console.log('I wish I was assigned to a variable :(');
}
Function length

Functions that are short are often easier to read. Splitting up your big functions into multiple smaller functions often helps. Try to keep function length between 10-15 lines.

Line Length

Limiting line length to about 80 characters will improve readability. Especially in complex statements where you want to remember what you were thinking in a few weeks from now (you do comment your code, right?).

Literal notation

Use the literal notation.

Right:

var myObject = {};
var myArray = [];
var myNumber = 1;
var myString = 'test';
var myBoolean = true;
Wrong:

var myObject = new Object();
var myArray = new Array();
var myNumber = new Number(1);
var myString = new String('test');
var myBoolean = new Boolean(1);
Loops

The body of every for in statement should be wrapped in an if statement that does filtering. If you're able to use $.each, use that instead.

Right:

for (name in object) {
    if (object.hasOwnProperty(name)) {
        ...
    }
}
Wrong:

for (name in object) {
    object[name];
}
Don't use functions within a loop.

Right:

var iterateItems = function(index, item) {
    debug.log('winning');
};

var items = ['item1', 'item2', 'item3', 'item4', 'item5'];

$.each(items, iterateItems);
Wrong:

var items = ['item1', 'item2', 'item3', 'item4', 'item5'];

$.each(items, function(index, item) {
    debug.log('losing');
});
Nested closures

Don't nest closures as it makes a mess out of your code.

Right:

var afterConnect = function() {
    debug.log('winning');
};

setTimeout(function() {
    client.connect(afterConnect);
}, 1000);
Wrong:

setTimeout(function() {
    client.connect(function() {
        debug.log('losing');
    });
}, 1000);
Object.freeze, Object.preventExtensions, Object.seal, with, eval

Evil! Please avoid these.

Object / Array creation

Put short declarations on a single line, and avoid using trailing commas.

Right:

var a = ['hello', 'world'];
var b = {
    'good': 'code',
    'is generally': 'pretty'
};
Wrong:

var a = [
    'hello', 'world'
];
var b = {"good": 'code'
        , is generally: 'pretty',
        };
Quotes

Use single quotes at all times, unless you are writing JSON.

Right:

var foo = 'bar';
Wrong:

var foo = "bar";
Return statements

Always try to return a function's value as soon as possible to avoid deep nesting of if-statements.

Right:

var isPercentage = function(val) {
    if (val < 0) {
        return false;
    }

    if (val > 100) {
        return false;
    }

    return true;
};
Wrong:

function isPercentage(val) {
   if (val >= 0) {
       if (val < 100) {
           return true;
       } else {
           return false;
       }
   } else {
       return false;
   }
}
In this example it's also fine to shorten things even further:

var isPercentage = function(val) {
    var isInRange = (val >= 0 && val <= 100);
    return isInRange;
};
Scope

Don't create globally scoped objects. We use RequireJS to handle our dependency management and this removes the need for global variables.

Semicolons

Following standard coding practices, semicolons are used where appropriate.

Tabs vs Spaces

4 spaces are used to indent lines of code. Any respectable editor will have the option of replacing tabs with spaces.

Ternary operations

Only use ternary operations where the benefits are substantial. If a ternary conditional requires sub-operations, then it should be an if statement for readability.

Right:

var howMuchCoffee = numPeople > 2 ? '12 cups' : '6 cups';
Wrong:

var shouldMakeMoreCoffee = numPeople > 2 && amountOfCoffeeLeft < 1 ? totalSleepiness < 2 || numPeople > 4 : numPeople < 4 && totalSleepiness > 1;
Trailing whitespace

Keeping code clean and simple starts with basics like whitespace. To keep your conscience clear have another look for extra spaces before committing your masterpiece.

Type checking

Never use typeof when there is a function that does the type checking for you. We include <%= link_to "jQuery", "http://jquery.com/", :class => "wl-regular-link" %> and <%= link_to "Underscore.js", "http://underscorejs.org/", :class => "wl-regular-link" %> which provide all the type checking you should need.

Right:

_.isString(someString)
$.isFunction(myFunction)
Wrong:

typeof someString === 'string'
typeof myFunction === 'function'
Variable and property names

Variables and properties should use lower camel case. It's also nice to have a variable that explains what it's used for. jQuery objects should be prepended with '$' for readability.

Right:

var adminUser = db.query('SELECT * FROM users ...');
var $commentBox = $('.comment_box', $rootel);
Wrong:

var admin_user = d.query('SELECT * FROM users ...');
var commentBox = $('.comment_box', rootel);
Variable declarations

Declare one variable per statement and consolidate them at the top of your file, this improves readability and maintenance of the code.

Right:

var keys = ['foo', 'bar'];
var values = [23, 42];

var object = {};
while (items.length) {
    var key = keys.pop();
    object[key] = values.pop();
}
Wrong:

var keys = ["foo", "bar"],
    values = [23, 42],
    object = {},
    key;

while (items.length) {
    key = keys.pop();
    object[key] = values.pop();
}
```